### ISSUE TITLE
PERF: Bring back <body is="polymer-body"> for fast rendering apps

### ISSUE THREAD

**Issue by ebidel** [Polymer/polymer#2654]
TL;DR Bring back a feature like `<body is="polymer-body">`. Related to this is https://github.com/Polymer/polymer/issues/2653. 

We used to have a feature `<body is="polymer-body">`, that gave you polymer features (bindings, `on-*` listeners, methods, etc.) and rendered things in light dom. `<x-app>` and `<template is="dom-bind">`'s limitation is that everything is hidden until imports are loaded/`dom-change` fires. All the perf improvements above are negated. **If we bring back `<body is="dom-bind"`>`, it would give a fast-path option towards super fast rendering apps but still benefit from Polymer features.**
# 
### Rendering bottlenecks

After _much_ research and experimenting in this area, I've found that the largest sources of first paint/load delay in most polymer apps are:
- Not using async scripts: `<script src="webcomponents-lite.js" async>` 
- Not using async imports: `<link rel="elements.html" async>`
- Using `<body unresolved>` (hide the page until polymer has declared ready)
- Using a single `<x-app>` or `<template is="dom-bind">` for all/most of the content. It's convenient to have the bindings and polymer features so we shouldn't get rid of it. For people that want fast apps, it would be nice to provide a fast alternative 

Other things that help:
- For larger apps, it's best to **not** vulcanize the main page. Vulcanizing index.html increases the weight of it. If you're using a lot of elements, that's a lot of inlined css/svg/html. On slow connections, this can be very noticeable b/c it slows the load of the main page.
- use `crisper --script-in-head` - this defer loads the js bundle at the top of the import bundle. For large vulcanized bundles, the parser has to go through a ton of svg/html/css before it starts loading the js. Therefore, elements upgrade later than they _could_, which delays the whole app. Ideally we could even move the script load into the main page, but there are timing issues when I've tried this. Example: `this.$.myid` may not resolve. This change alone can [speed up the vulcanized script download by 80%]
### What fast can look like

Traditional polymer app: https://ebidel.github.io/polymer-experiments/polymersummit/fouc/unresolved.html is not fast on a slow connection. It async's nothing and uses `<body unresolved>`.

Hotness: https://ebidel.github.io/polymer-experiments/polymersummit/fouc/appshell.html is super fast. **It paints instantly, even on a 2g connection**! It uses all the patterns above but does not use Polymer features like bindings (just consumes elements). 

Note: WPT tends to lie! I no longer trust their vms/setup. It reports 8.6s render on chrome desktop 2g ([results](http://www.webpagetest.org/result/151031_MK_J2/). Chrome devtools w/ 2g emulation reports 2s fp.

<img width="1201" alt="screen shot 2015-10-30 at 5 03 39 pm" src="https://cloud.githubusercontent.com/assets/238208/10860667/f9c5b61e-7f28-11e5-8382-1d87bed8366f.png">

For the best of both worlds (fast rendering + leveraging polymer features), `<body is="polymer-body">` would be killer.
### This can be faster than other frameworks

The browser is **REALLY** fast at rendering markup. We shouldn't hide it. This is a huge advantage to any other framework. We work directly in markup and elements upgrades can be used to our advantage. Other frameworks generate markup. Let's take advantage of the rendering pipeline we're given.

**Comment by ebidel** [comment]
The tradeoff here is visual FOUC of `{{}}` showing before the element is upgraded. This is something users will have to deal with if they chose the fast path, but is easily dealt with with CSS. I'd love to have a POC element that I can use to use show how. Tried looking at `dom-bind` myself, but there's a lot going on.

**Comment by robdodson** [comment]
@ebidel could we make a special attribute binding for textContent? That way the server can send down static text inside the tags instead of `{{}}` and then <my-section data-text="{{article}}"> perhaps gets fulfilled later?

**Comment by ebidel** [comment]
That's an interesting idea Rob. It changes the semantics of the element though. In some cases, users won't be able to change the internals of an element (e.g. consuming elements).

**Comment by ebidel** [comment]
Re https://github.com/Polymer/polymer/issues/2654#issuecomment-152631777, you can do `<my-section text-content="{{article}}">`

**Comment by ebidel** [comment]
@tjsavage @kevinpschaaf can we include this in Polymer.next planning?

**Comment by ebidel** [comment]
This issue is meant to provide (bring back) a solution for that pattern. Do we have a polymer 2 vended element for light dom and bindings? The summit site is using one off elements :/

**Comment by TimvdLippe** [comment]
Oh I see. The request is to have bindings in the light dom and then have a Polymer mixin that parses the bindings in the lightdom next to parsing in its templates. I am not sure how that would work, as eagerly setting bindings in a non-template (e.g. not inert) will eagerly set the bindings. That means that attributes and textcontent will probably be corrupt until Polymer fixes them.

I am not familiar with the old solution, so maybe @kevinpschaaf or @sorvell have any thoughts on this issue. Thanks for clarifying @ebidel !

**Comment by ebidel** [comment]
Right you are :) https://github.com/Polymer/polymer/issues/2654#issuecomment-152625607

I think users that opt in to a fast path like this would know how to control it. But we should give them the tools to make their site fast!
